# 1. Data Extraction

In this notebook we will extract our data and analyse it. For that purpose, we are importing from our library where we define the
```bcrp_dataframe``` dataframe. This function will allows us to use the API interface of the Central Bank of Reserve of Peru (BCRP) to automatically create a pandas dataframe with the necessary codes.

## 1.1 Libraries

We import the necessary libraries, including our own library in the modules file

In [3]:
# Warnings
import warnings
warnings.filterwarnings("ignore")

# Basic Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import requests
import seaborn as sns
from scipy import stats
from functools import reduce

# Statsmodels
import statsmodels.api as sm
import pmdarima as pmd
from pmdarima.arima import auto_arima
from statsmodels.tsa.api import VAR
from statsmodels.tsa.vector_ar.var_model import VARResults
from statsmodels.tsa.statespace.sarimax import SARIMAX
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.tsa.stattools import adfuller
from statsmodels.tsa.seasonal import STL

# Machine Learning models
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV, TimeSeriesSplit
from sklearn.linear_model import Ridge, Lasso, ElasticNet, ElasticNetCV, LinearRegression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    mean_absolute_percentage_error,
    median_absolute_error,
    r2_score,
    precision_score

)

from xgboost import XGBRegressor



In [4]:
# We import our own functions
import sys
sys.path.append('../../..')  # Move two levels up to the project root
from modules.functions import *

## 1.2 Extraction
We define our inputs and apply them the ```bcrp_dataframe``` function in order to obtain the pandas dataframe with the corresponding series

We define the following inputs:

    series     = the code of the series we are going to extract
    start_date = the starting date, when the BCRP starts using the interest rate as a policy measure
    end_date   = December 2019
    freq       = Monthly frequency

### df_1
We can now create the first dataframe with the ```bcrp_dataframe``` function. This dataframe contains out target variable Headline Inflation 

In [5]:
series     = ['PN01271PM']
start_date = '2003-09'
end_date   = '2019-12'
freq       = 'Mensual'

In [6]:
df_1 = bcrp_dataframe( series , start_date , end_date , freq )
df_1.head()

Índice de precios Lima Metropolitana (var% mensual) - IPC
Fecha                                                                
2003-09-01                                           0.558598        
2003-10-01                                           0.049032        
2003-11-01                                           0.167685        
2003-12-01                                           0.563951        
2004-01-01                                           0.537447

In [7]:
df_1 = get_trend(df_1)
df_1.head()

Índice de precios Lima Metropolitana (var% mensual) - IPC
Fecha                                                                
2003-09-01                                           0.518834        
2003-10-01                                           0.492247        
2003-11-01                                           0.465462        
2003-12-01                                           0.438413        
2004-01-01                                           0.411138

### df_2
We create the second dataframe with the ```bcrp_dataframe``` function. This dataframe contains rate variables. We use this variables in levels. It is not necessary to differentiate them.

In [8]:
series     = ['PD04722MM']
start_date = '2003-09'
end_date   = '2019-12'
freq       = 'Mensual'

In [9]:
df_2 = bcrp_dataframe( series , start_date , end_date , freq )
df_2.head()

Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria
Fecha                                                                                                   
2003-09-01                                               2.75                                           
2003-10-01                                               2.75                                           
2003-11-01                                               2.50                                           
2003-12-01                                               2.50                                           
2004-01-01                                               2.50

### df_3
We create the third dataframe with the ```bcrp_dataframe``` function. This dataframe contains monetary variables as well as commodities. We differentiate those variables in order to get their monthly % change

In [10]:
series     = ['PN00495MM', 'PN06481IM', 'PN02125PM', 'PN01661XM','PN01662XM','PN01664XM','PN01660XM']
start_date = '2003-09'
end_date   = '2019-12'
freq       = 'Mensual'

In [13]:
df_3 = bcrp_dataframe( series , start_date , end_date , freq )
df_3 = np.log(df_3)
df_3 = df_3.dropna()
df_3.head()

Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado  \
Fecha                                                                                        
2003-09-01                                           8.591877                                
2003-10-01                                           8.602051                                
2003-11-01                                           8.613430                                
2003-12-01                                           8.629901                                
2004-01-01                                           8.649204                                

            Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)  \
Fecha                                                                                               
2003-09-01                                           9.185581                                       
2003-10-01                                           9.191108                                       
2003-11-01                                           9.240173                                       
2003-12-01                                           9.229584                                       
2004-01-01                                           9.265177                                       

            Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)  \
Fecha                                                                                    
2003-09-01                                           5.440117                            
2003-10-01                                           5.495508                            
2003-11-01                                           5.493832                            
2003-12-01                                           5.488208                            
2004-01-01                                           5.482848                            

            Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por toneladas)  \
Fecha                                                                                             
2003-09-01                                           4.884206                                     
2003-10-01                                           4.900658                                     
2003-11-01                                           5.009696                                     
2003-12-01                                           5.046169                                     
2004-01-01                                           5.049709                                     

            Cotizaciones de productos (promedio del periodo) - Maíz - EEUU (US$ por toneladas)  \
Fecha                                                                                            
2003-09-01                                           4.419001                                    
2003-10-01                                           4.390800                                    
2003-11-01                                           4.463461                                    
2003-12-01                                           4.511148                                    
2004-01-01                                           4.571827                                    

            Cotizaciones de productos (promedio del periodo) - Aceite Soya - EEUU (US$ por toneladas)  \
Fecha                                                                                                   
2003-09-01                                           6.238332                                           
2003-10-01                                           6.404890                                           
2003-11-01                                           6.421768                                           
2003-12-01                                           6.471752                                           
200

In [14]:
df = df_1.join(df_2).join(df_3)
df.dropna(inplace=True)
df.head()

Índice de precios Lima Metropolitana (var% mensual) - IPC  \
Fecha                                                                   
2003-09-01                                           0.518834           
2003-10-01                                           0.492247           
2003-11-01                                           0.465462           
2003-12-01                                           0.438413           
2004-01-01                                           0.411138           

            Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria  \
Fecha                                                                                                      
2003-09-01                                               2.75                                              
2003-10-01                                               2.75                                              
2003-11-01                                               2.50                                              
2003-12-01                                               2.50                                              
2004-01-01                                               2.50                                              

            Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado  \
Fecha                                                                                        
2003-09-01                                           8.591877                                
2003-10-01                                           8.602051                                
2003-11-01                                           8.613430                                
2003-12-01                                           8.629901                                
2004-01-01                                           8.649204                                

            Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)  \
Fecha                                                                                               
2003-09-01                                           9.185581                                       
2003-10-01                                           9.191108                                       
2003-11-01                                           9.240173                                       
2003-12-01                                           9.229584                                       
2004-01-01                                           9.265177                                       

            Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)  \
Fecha                                                                                    
2003-09-01                                           5.440117                            
2003-10-01                                           5.495508                            
2003-11-01                                           5.493832                            
2003-12-01                                           5.488208                            
2004-01-01                                           5.482848                            

            Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por toneladas)  \
Fecha                                                                                             
2003-09-01                                           4.884206                                     
2003-10-01                                           4.900658                                     
2003-11-01                                           5.009696                                     
2003-12-01                                           5.046169                                     
2004-01-01                                           5.049709                                     

            Cotizaciones de productos (promedio del

In [15]:
df.tail()

Índice de precios Lima Metropolitana (var% mensual) - IPC  \
Fecha                                                                   
2019-08-01                                           0.155238           
2019-09-01                                           0.151966           
2019-10-01                                           0.148757           
2019-11-01                                           0.145714           
2019-12-01                                           0.142886           

            Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria  \
Fecha                                                                                                      
2019-08-01                                               2.50                                              
2019-09-01                                               2.50                                              
2019-10-01                                               2.50                                              
2019-11-01                                               2.25                                              
2019-12-01                                               2.25                                              

            Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado  \
Fecha                                                                                        
2019-08-01                                          10.783622                                
2019-09-01                                          10.792065                                
2019-10-01                                          10.795930                                
2019-11-01                                          10.802030                                
2019-12-01                                          10.813291                                

            Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)  \
Fecha                                                                                               
2019-08-01                                          11.131211                                       
2019-09-01                                          11.125202                                       
2019-10-01                                          11.127057                                       
2019-11-01                                          11.109024                                       
2019-12-01                                          11.131899                                       

            Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)  \
Fecha                                                                                    
2019-08-01                                           5.748980                            
2019-09-01                                           5.748917                            
2019-10-01                                           5.747810                            
2019-11-01                                           5.746721                            
2019-12-01                                           5.744579                            

            Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por toneladas)  \
Fecha                                                                                             
2019-08-01                                           5.194315                                     
2019-09-01                                           5.165015                                     
2019-10-01                                           5.204847                                     
2019-11-01                                           5.274724                                     
2019-12-01                                           5.306730                                     

            Cotizaciones de productos (promedio del

## 1.3 Data Inspection
We inspect the df. We first verify that all values are non-null. The, we apply the ```describe``` function to see the main variables.

In [16]:
df.isna().sum()

Índice de precios Lima Metropolitana (var% mensual) - IPC                                       0
Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria    0
Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado                  0
Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)           0
Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)                      0
Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por toneladas)             0
Cotizaciones de productos (promedio del periodo) - Maíz - EEUU (US$ por toneladas)              0
Cotizaciones de productos (promedio del periodo) - Aceite Soya - EEUU (US$ por toneladas)       0
Cotizaciones de productos (promedio del periodo) - Petróleo - WTI (US$ por barriles)            0
dtype: int64

In [17]:
df.describe()

Índice de precios Lima Metropolitana (var% mensual) - IPC  \
count                                         196.000000           
mean                                            0.237890           
std                                             0.106886           
min                                             0.056201           
25%                                             0.156165           
50%                                             0.241131           
75%                                             0.293527           
max                                             0.518834           

       Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria  \
count                                         196.000000                                              
mean                                            3.665816                                              
std                                             1.106475                                              
min                                             1.250000                                              
25%                                             2.937500                                              
50%                                             3.750000                                              
75%                                             4.250000                                              
max                                             6.500000                                              

       Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado  \
count                                         196.000000                                
mean                                            9.967408                                
std                                             0.659690                                
min                                             8.591877                                
25%                                             9.449695                                
50%                                            10.131215                                
75%                                            10.547071                                
max                                            10.813291                                

       Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)  \
count                                         196.000000                                       
mean                                           10.529348                                       
std                                             0.629997                                       
min                                             9.185581                                       
25%                                            10.113836                                       
50%                                            10.794583                                       
75%                                            11.040132                                       
max                                            11.131899                                       

       Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)  \
count                                         196.000000                            
mean                                            5.597523                            
std                                             0.106667                            
min                                             5.439158                            
25%                                             5.502205                            
50%                                             5.606932                            
75%                                             5.695481                            
max                                             5.777829                            

       Cotizaciones de productos 

We have 195 observation ranging from ```2003-10-01``` to ```2019-12-01```. The mean of monthly % change of all CPI variables is around 0.2. The mean of the lacing rate and the interest rate is 10.7% and 3.67%, respectively. The three monetary variables have a small monthly % change, around 0.01 and 0.001 for Minimum Wage index. 

## 1.4 Data adjustment
We will rename the columns for easier identification of the variables. We will also create a new dataframe with the lags of the variables. 

In [18]:
# New column names
columns = {
    'Índice de precios Lima Metropolitana (var% mensual) - IPC': 'CPI',
    'Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria': 'Monetary Policy Rate',
    'Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado': 'Circulating Currency Seasonally Adjusted (mill S/)',
    'Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)': 'Net International Reserves (mill $)',
    'Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)': 'Real Minimum Wage (Index)',
    'Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por toneladas)': 'Wheat (US$ per ton)',
    'Cotizaciones de productos (promedio del periodo) - Maíz - EEUU (US$ por toneladas)': 'Corn  (US$ per ton)',
    'Cotizaciones de productos (promedio del periodo) - Aceite Soya - EEUU (US$ por toneladas)': 'Soybean oil (US$ per ton)',
    'Cotizaciones de productos (promedio del periodo) - Petróleo - WTI (US$ por barriles)': 'Crude oil (US$ per barrel)'  
}

# We rename the columns so they are easier to analyse
df.rename(columns=columns, inplace=True)

In [19]:
df_lags = df.copy()

for variable in df_lags.columns[1:]:
    df_lags[f'{variable}_lag_1'] = df_lags[variable].shift()
    df_lags[f'{variable}_lag_2'] = df_lags[variable].shift(2)
    df_lags[f'{variable}_lag_3'] = df_lags[variable].shift(3)
    df_lags[f'{variable}_lag_4'] = df_lags[variable].shift(4)

In [20]:
# We delete contemporary variables
df_lags.drop(columns = ['Monetary Policy Rate','Circulating Currency Seasonally Adjusted (mill S/)',
       'Net International Reserves (mill $)', 'Real Minimum Wage (Index)', 'Wheat (US$ per ton)', 'Corn  (US$ per ton)', 
       'Soybean oil (US$ per ton)', 'Crude oil (US$ per barrel)'], inplace = True)

df_lags = df_lags.dropna()

## 1.5 Save Results
We save it to the ```input``` folder, where we can use it to do the forecasting in the next notebook.

In [21]:
df.to_csv('../../../input/df_raw_test.csv')

In [22]:
df_lags.to_csv('../../../input/df_lags_test.csv')

In [23]:
df_lags.tail()

CPI  Monetary Policy Rate_lag_1  Monetary Policy Rate_lag_2  \
Fecha                                                                          
2019-08-01  0.155238                        2.75                        2.75   
2019-09-01  0.151966                        2.50                        2.75   
2019-10-01  0.148757                        2.50                        2.50   
2019-11-01  0.145714                        2.50                        2.50   
2019-12-01  0.142886                        2.25                        2.50   

            Monetary Policy Rate_lag_3  Monetary Policy Rate_lag_4  \
Fecha                                                                
2019-08-01                        2.75                        2.75   
2019-09-01                        2.75                        2.75   
2019-10-01                        2.75                        2.75   
2019-11-01                        2.50                        2.75   
2019-12-01                        2.50                        2.50   

            Circulating Currency Seasonally Adjusted (mill S/)_lag_1  \
Fecha                                                                  
2019-08-01                                          10.775826          
2019-09-01                                          10.783622          
2019-10-01                                          10.792065          
2019-11-01                                          10.795930          
2019-12-01                                          10.802030          

            Circulating Currency Seasonally Adjusted (mill S/)_lag_2  \
Fecha                                                                  
2019-08-01                                          10.772159          
2019-09-01                                          10.775826          
2019-10-01                                          10.783622          
2019-11-01                                          10.792065          
2019-12-01                                          10.795930          

            Circulating Currency Seasonally Adjusted (mill S/)_lag_3  \
Fecha                                                                  
2019-08-01                                          10.761742          
2019-09-01                                          10.772159          
2019-10-01                                          10.775826          
2019-11-01                                          10.783622          
2019-12-01                                          10.792065          

            Circulating Currency Seasonally Adjusted (mill S/)_lag_4  \
Fecha                                                                  
2019-08-01                                          10.762972          
2019-09-01                                          10.761742          
2019-10-01                                          10.772159          
2019-11-01                                          10.775826          
2019-12-01                                          10.783622          

            Net International Reserves (mill $)_lag_1  ...  \
Fecha                                                  ...   
2019-08-01                                  11.108755  ...   
2019-09-01                                  11.131211  ...   
2019-10-01                                  11.125202  ...   
2019-11-01                                  11.127057  ...   
2019-12-01                                  11.109024  ...   

            Corn  (US$ per ton)_lag_3  Corn  (US$ per ton)_lag_4  \
Fecha                                                              
2019-08-01                   4.928328                   4.861634   
2019-09-01                   5.083703                   4.928328   
2019-10-01                   5.095125                   5.083703   
2019-11-01                   4.959057                   5.095125   
2019-12-01                   4.913953                   4.959057   

            Soybean oil (US$ per 

In [25]:
df.tail()

CPI  Monetary Policy Rate  \
Fecha                                        
2019-08-01  0.155238                  2.50   
2019-09-01  0.151966                  2.50   
2019-10-01  0.148757                  2.50   
2019-11-01  0.145714                  2.25   
2019-12-01  0.142886                  2.25   

            Circulating Currency Seasonally Adjusted (mill S/)  \
Fecha                                                            
2019-08-01                                          10.783622    
2019-09-01                                          10.792065    
2019-10-01                                          10.795930    
2019-11-01                                          10.802030    
2019-12-01                                          10.813291    

            Net International Reserves (mill $)  Real Minimum Wage (Index)  \
Fecha                                                                        
2019-08-01                            11.131211                   5.748980   
2019-09-01                            11.125202                   5.748917   
2019-10-01                            11.127057                   5.747810   
2019-11-01                            11.109024                   5.746721   
2019-12-01                            11.131899                   5.744579   

            Wheat (US$ per ton)  Corn  (US$ per ton)  \
Fecha                                                  
2019-08-01             5.194315             4.959057   
2019-09-01             5.165015             4.913953   
2019-10-01             5.204847             4.978704   
2019-11-01             5.274724             4.942720   
2019-12-01             5.306730             4.950479   

            Soybean oil (US$ per ton)  Crude oil (US$ per barrel)  
Fecha                                                              
2019-08-01                   6.443189                    4.003798  
2019-09-01                   6.462062                    4.040550  
2019-10-01                   6.503629                    3.988299  
2019-11-01                   6.529491                    4.044437  
2019-12-01                   6.585121                    4.092312